File: Docs/examples/python/advanced_usage.py
"""
Advanced GRID API Usage Examples

Demonstrates advanced features including:
- Positions management
- Price history and charting
- Instruments discovery
- Transfers between accounts
- Market statistics
"""

import time
from typing import List, Dict, Optional
from datetime import datetime, timedelta
from collections import defaultdict

from trading_client import GridTradingClient


class AdvancedGridClient:
    """Advanced GRID API client with helper methods"""
    
    def __init__(self, private_key: str, public_key: str):
        self.client = GridTradingClient(private_key, public_key)
    
    def get_portfolio_overview(self):
        """Get complete portfolio overview"""
        print('=== Portfolio Overview ===\n')
        
        # Get positions
        positions = self.client.get_positions(status='open')
        
        print(f'Open Positions: {len(positions)}')
        total_value = 0
        
        for pos in positions:
            value = float(pos.get('current_market_value', 0))
            cost = float(pos['total_cost'])
            pnl = value - cost
            pnl_pct = (pnl / cost) * 100
            
            print(f"  {pos['instrument_name']}:")
            print(f"    Quantity: {pos['quantity']} units")
            print(f"    Avg Cost: ${pos['average_cost']}")
            print(f"    Current Value: ${value:.2f}")
            print(f"    P&L: ${pnl:.2f} ({pnl_pct:+.2f}%)")
            
            total_value += value
        
        print(f'\nTotal Portfolio Value: ${total_value:.2f}\n')
        
        # Get account balances
        accounts = self.client.get_trading_accounts()
        
        print('Account Balances:')
        for acct in accounts:
            symbol = acct.get('instrument_symbol', acct.get('currency', 'Unknown'))
            print(f"  {symbol}:")
            print(f"    Available: {acct.get('available_balance', acct.get('available', 0))}")
            print(f"    Locked: {acct.get('locked_balance', acct.get('reserved', 0))}")
            print(f"    Total: {acct.get('total_balance', acct.get('total', 0))}")
        
        # Get consumption balance
        consumption = self.client.get_consumption_instruments()
        
        if consumption:
            print('\nConsumption Balances:')
            for inst in consumption:
                used_pct = (inst['tokens_used'] / inst['tokens_purchased']) * 100
                print(f"  {inst['instrument_name']}:")
                print(f"    Units: {inst['total_amount']} ({inst['tradeable_amount']} tradeable)")
                print(f"    Tokens: {inst['tokens_used']:,} / {inst['tokens_purchased']:,}")
                print(f"    Usage: {used_pct:.1f}%")
    
    def analyze_market(self, market_id: str, days: int = 30):
        """Analyze market with price history"""
        print(f'\n=== Market Analysis: {market_id} ===\n')
        
        # Get market stats
        stats = self.client.get_market_stats(market_id)
        
        print('24h Statistics:')
        print(f"  Current Price: ${stats['current_price']}")
        print(f"  24h Change: ${stats['price_change_24h']} ({stats['price_change_24h_percent']}%)")
        print(f"  24h High: ${stats['high_24h']}")
        print(f"  24h Low: ${stats['low_24h']}")
        print(f"  24h Volume: {stats['volume_24h']} units (${stats['volume_24h_value']})")
        
        # Get price history
        now = int(time.time())
        start = now - (days * 24 * 60 * 60)
        
        candles = self.client.get_price_history(market_id, '1d', start, now)
        
        print(f'\nPrice History ({days} days):')
        print(f'  Candles: {len(candles)}')
        
        if candles:
            first_candle = candles[0]
            last_candle = candles[-1]
            
            price_change = float(last_candle['close']) - float(first_candle['open'])
            price_change_pct = (price_change / float(first_candle['open'])) * 100
            
            print(f"  Period Start: ${first_candle['open']}")
            print(f"  Period End: ${last_candle['close']}")
            print(f"  Period Change: ${price_change:.2f} ({price_change_pct:+.2f}%)")
            
            # Calculate volatility
            closes = [float(c['close']) for c in candles]
            returns = [(closes[i] - closes[i-1]) / closes[i-1] for i in range(1, len(closes))]
            
            avg_return = sum(returns) / len(returns)
            variance = sum((r - avg_return) ** 2 for r in returns) / len(returns)
            volatility = (variance ** 0.5) * 100
            
            print(f"  Volatility: {volatility:.2f}%")
            
            # Total volume
            total_volume = sum(c['volume'] for c in candles)
            print(f"  Total Volume: {total_volume} units")
    
    def find_best_ai_models(
        self,
        min_context_window: int = 100000,
        max_price: float = 100
    ) -> List[Dict]:
        """Find best AI commodity instruments"""
        print('\n=== AI Model Search ===\n')
        print(f'Criteria: Context >= {min_context_window:,}, Price <= ${max_price}\n')
        
        instruments = self.client.list_instruments()
        ai_instruments = [i for i in instruments if i['instrument_type'] == 'ai_commodity']
        
        qualified = []
        
        for inst in ai_instruments:
            details = self.client.get_instrument(inst['instrument_id'])
            
            if not details.get('ai_specs') or not details['ai_specs'].get('context_window'):
                continue
            
            if details['ai_specs']['context_window'] >= min_context_window:
                price = float(details['last_trade_price']) if details['last_trade_price'] else None
                
                if not price or price <= max_price:
                    qualified.append({
                        'symbol': details['symbol'],
                        'instrument_id': details['instrument_id'],
                        'context_window': details['ai_specs']['context_window'],
                        'throughput': details['ai_specs']['token_throughput'],
                        'price': price or 'N/A',
                        'models': details['ai_specs'].get('qualifying_models', [])
                    })
        
        # Sort by context window (descending)
        qualified.sort(key=lambda x: x['context_window'], reverse=True)
        
        print(f'Found {len(qualified)} qualifying instruments:\n')
        
        for inst in qualified:
            print(f"{inst['symbol']}:")
            print(f"  Context Window: {inst['context_window']:,} tokens")
            print(f"  Throughput: {inst['throughput']} tokens/sec")
            print(f"  Price: ${inst['price']:.2f}" if isinstance(inst['price'], (int, float)) else f"  Price: {inst['price']}")
            print(f"  Models: {', '.join(inst['models']) if inst['models'] else 'None listed'}")
            print('')
        
        return qualified
    
    def buy_for_ai_consumption(
        self,
        market_id: str,
        instrument_id: str,
        quantity: int
    ):
        """Complete workflow: Buy and prepare for AI consumption"""
        print('\n=== Buy and Prepare for AI Consumption ===\n')
        
        # Step 1: Check consumption balance
        consumption = self.client.get_consumption_instruments()
        existing = next((i for i in consumption if i['instrument_id'] == instrument_id), None)
        
        if existing and existing['tradeable_amount'] >= quantity:
            print(f"✓ Already have {existing['tradeable_amount']} units in consumption")
            return
        
        # Step 2: Check trading balance
        trading_accounts = self.client.get_trading_accounts()
        trading_acct = next((a for a in trading_accounts if a.get('instrument_id') == instrument_id), None)
        
        available_balance = int(trading_acct['available_balance']) if trading_acct else 0
        
        if available_balance < quantity:
            print(f'Insufficient trading balance. Need {quantity}, have {available_balance}')
            print('Placing buy order...')
            
            # Get current market price
            ticker = self.client.get_ticker(market_id)
            buy_price = float(ticker.get('lowest_ask', ticker.get('last_price', 0)))
            
            # Place limit order at current ask
            order = self.client.place_order(
                market_id=market_id,
                side='buy',
                order_type='limit',
                quantity=str(quantity),
                price=f'{buy_price:.2f}',
                time_in_force='gtc'
            )
            
            print(f"✓ Order placed: {order['order_id']}")
            print('  Waiting for fill...')
            
            # In production, use WebSocket to wait for fill
            time.sleep(5)
        else:
            print(f'✓ Sufficient trading balance: {available_balance} units')
        
        # Step 3: Transfer to consumption
        print(f'Transferring {quantity} units to consumption...')
        
        transfer = self.client.transfer_to_consumption(instrument_id, quantity)
        print(f"✓ Transfer initiated: {transfer['transfer_id']}")
        
        # Step 4: Confirm transfer
        print('Waiting for transfer completion...')
        
        completed = False
        attempts = 0
        
        while not completed and attempts < 10:
            time.sleep(2)
            
            history = self.client.get_transfer_history(limit=100)
            transfer_record = next((t for t in history if t['transfer_id'] == transfer['transfer_id']), None)
            
            if transfer_record:
                print('✓ Transfer completed')
                completed = True
            
            attempts += 1
        
        if not completed:
            print('⚠ Transfer taking longer than expected. Check transfer history.')
        
        print('\n✓ Units ready for AI consumption!')
    
    def analyze_transfer_patterns(self, market_id: str, days: int = 30):
        """Analyze transfer patterns over time"""
        print(f'\n=== Transfer Pattern Analysis ({days} days) ===\n')
        
        history = self.client.get_transfer_history(market_id=market_id)
        
        # Filter to last N days
        cutoff = datetime.now() - timedelta(days=days)
        recent = [
            t for t in history
            if datetime.fromisoformat(t['transferred_at'].replace('Z', '+00:00')) > cutoff
        ]
        
        # Analyze by direction
        to_consumption = [t for t in recent if 'consumption' in t['account_id']]
        to_trading = [t for t in recent if 'trading' in t['account_id']]
        
        # Calculate totals by instrument
        instrument_totals = defaultdict(lambda: {'to_consumption': 0, 'to_trading': 0})
        
        for t in to_consumption:
            instrument_totals[t['instrument_id']]['to_consumption'] += t['quantity']
        
        for t in to_trading:
            instrument_totals[t['instrument_id']]['to_trading'] += t['quantity']
        
        # Report
        print(f"Total Transfers: {len(recent)}")
        print(f"  To Consumption: {len(to_consumption)}")
        print(f"  To Trading: {len(to_trading)}")
        
        print('\nBy Instrument:')
        for inst_id, totals in instrument_totals.items():
            print(f"  {inst_id}:")
            print(f"    → Consumption: {totals['to_consumption']} units")
            print(f"    → Trading: {totals['to_trading']} units")
            net = totals['to_consumption'] - totals['to_trading']
            print(f"    Net Flow: {net:+d} units")


if __name__ == '__main__':
    # Load keys from files
    with open('./ed25519.key', 'r') as f:
        private_key = f.read().strip()
    
    with open('./ed25519_pub.der', 'r') as f:
        public_key = f.read().strip()
    
    # Initialize client
    client = AdvancedGridClient(private_key, public_key)
    
    try:
        # Get portfolio overview
        client.get_portfolio_overview()
        
        # Analyze market
        markets = client.client.get_markets()
        if markets:
            market_id = markets[0]['market_id']
            
            client.analyze_market(market_id, days=30)
            
            # Find best AI models
            best_models = client.find_best_ai_models(
                min_context_window=100000,
                max_price=100
            )
            
            if best_models:
                print(f'\nRecommended: {best_models[0]["symbol"]}')
            
            # Analyze transfer patterns
            client.analyze_transfer_patterns(market_id, days=7)
        
    except Exception as e:
        print(f'Error: {e}')

File: Docs/examples/javascript/advanced-usage.js
/**
 * Advanced GRID API Usage Examples
 * 
 * Demonstrates advanced features including:
 * - Positions management
 * - Price history and charting
 * - Instruments discovery
 * - Transfers between accounts
 * - Market statistics
 */

const axios = require('axios');
const { SignatureAuth } = require('./auth');
const { GridTradingClient } = require('./trading-client');

class AdvancedGridClient {
  constructor(privateKey, publicKey) {
    this.client = new GridTradingClient(privateKey, publicKey);
  }

  /**
   * Get complete portfolio overview
   */
  async getPortfolioOverview() {
    console.log('=== Portfolio Overview ===\n');
    
    // Get positions
    const positions = await this.client.getPositions({ status: 'open' });
    
    console.log(`Open Positions: ${positions.length}`);
    let totalValue = 0;
    
    for (const pos of positions) {
      const value = parseFloat(pos.current_market_value) || 0;
      const cost = parseFloat(pos.total_cost);
      const pnl = value - cost;
      const pnlPct = (pnl / cost) * 100;
      
      console.log(`  ${pos.instrument_name}:`);
      console.log(`    Quantity: ${pos.quantity} units`);
      console.log(`    Avg Cost: $${pos.average_cost}`);
      console.log(`    Current Value: $${value.toFixed(2)}`);
      console.log(`    P&L: $${pnl.toFixed(2)} (${pnlPct.toFixed(2)}%)`);
      
      totalValue += value;
    }
    
    console.log(`\nTotal Portfolio Value: $${totalValue.toFixed(2)}\n`);
    
    // Get account balances
    const accounts = await this.client.getTradingAccounts();
    
    console.log('Account Balances:');
    for (const acct of accounts) {
      console.log(`  ${acct.instrument_symbol || acct.currency}:`);
      console.log(`    Available: ${acct.available_balance}`);
      console.log(`    Locked: ${acct.locked_balance || acct.reserved}`);
      console.log(`    Total: ${acct.total_balance || acct.total}`);
    }
    
    // Get consumption balance
    const consumption = await this.client.getConsumptionInstruments();
    
    if (consumption.length > 0) {
      console.log('\nConsumption Balances:');
      for (const inst of consumption) {
        const usedPct = (inst.tokens_used / inst.tokens_purchased) * 100;
        console.log(`  ${inst.instrument_name}:`);
        console.log(`    Units: ${inst.total_amount} (${inst.tradeable_amount} tradeable)`);
        console.log(`    Tokens: ${inst.tokens_used.toLocaleString()} / ${inst.tokens_purchased.toLocaleString()}`);
        console.log(`    Usage: ${usedPct.toFixed(1)}%`);
      }
    }
  }

  /**
   * Analyze market with price history
   */
  async analyzeMarket(marketId, days = 30) {
    console.log(`\n=== Market Analysis: ${marketId} ===\n`);
    
    // Get market stats
    const stats = await this.client.getMarketStats(marketId);
    
    console.log('24h Statistics:');
    console.log(`  Current Price: $${stats.current_price}`);
    console.log(`  24h Change: $${stats.price_change_24h} (${stats.price_change_24h_percent}%)`);
    console.log(`  24h High: $${stats.high_24h}`);
    console.log(`  24h Low: $${stats.low_24h}`);
    console.log(`  24h Volume: ${stats.volume_24h} units ($${stats.volume_24h_value})`);
    
    // Get price history
    const now = Math.floor(Date.now() / 1000);
    const start = now - (days * 24 * 60 * 60);
    
    const candles = await this.client.getPriceHistory(marketId, '1d', start, now);
    
    console.log(`\nPrice History (${days} days):`);
    console.log(`  Candles: ${candles.length}`);
    
    if (candles.length > 0) {
      const firstCandle = candles[0];
      const lastCandle = candles[candles.length - 1];
      
      const priceChange = parseFloat(lastCandle.close) - parseFloat(firstCandle.open);
      const priceChangePct = (priceChange / parseFloat(firstCandle.open)) * 100;
      
      console.log(`  Period Start: $${firstCandle.open}`);
      console.log(`  Period End: $${lastCandle.close}`);
      console.log(`  Period Change: $${priceChange.toFixed(2)} (${priceChangePct.toFixed(2)}%)`);
      
      // Calculate volatility
      const closes = candles.map(c => parseFloat(c.close));
      const returns = [];
      for (let i = 1; i < closes.length; i++) {
        const ret = (closes[i] - closes[i-1]) / closes[i-1];
        returns.push(ret);
      }
      
      const avgReturn = returns.reduce((a, b) => a + b, 0) / returns.length;
      const variance = returns.reduce((sum, ret) => sum + Math.pow(ret - avgReturn, 2), 0) / returns.length;
      const volatility = Math.sqrt(variance) * 100;
      
      console.log(`  Volatility: ${volatility.toFixed(2)}%`);
      
      // Calculate total volume
      const totalVolume = candles.reduce((sum, c) => sum + c.volume, 0);
      console.log(`  Total Volume: ${totalVolume} units`);
    }
  }

  /**
   * Find best AI commodity instruments
   */
  async findBestAIModels(minContextWindow = 100000, maxPrice = 100) {
    console.log('\n=== AI Model Search ===\n');
    console.log(`Criteria: Context >= ${minContextWindow}, Price <= $${maxPrice}\n`);
    
    const instruments = await this.client.listInstruments();
    const aiInstruments = instruments.filter(i => i.instrument_type === 'ai_commodity');
    
    const qualified = [];
    
    for (const inst of aiInstruments) {
      const details = await this.client.getInstrument(inst.instrument_id);
      
      if (!details.ai_specs || !details.ai_specs.context_window) continue;
      
      if (details.ai_specs.context_window >= minContextWindow) {
        const price = details.last_trade_price ? parseFloat(details.last_trade_price) : null;
        
        if (!price || price <= maxPrice) {
          qualified.push({
            symbol: details.symbol,
            instrumentId: details.instrument_id,
            contextWindow: details.ai_specs.context_window,
            throughput: details.ai_specs.token_throughput,
            price: price || 'N/A',
            models: details.ai_specs.qualifying_models || []
          });
        }
      }
    }
    
    // Sort by context window (descending)
    qualified.sort((a, b) => b.contextWindow - a.contextWindow);
    
    console.log(`Found ${qualified.length} qualifying instruments:\n`);
    
    qualified.forEach(inst => {
      console.log(`${inst.symbol}:`);
      console.log(`  Context Window: ${inst.contextWindow.toLocaleString()} tokens`);
      console.log(`  Throughput: ${inst.throughput} tokens/sec`);
      console.log(`  Price: ${typeof inst.price === 'number' ? '$' + inst.price.toFixed(2) : inst.price}`);
      console.log(`  Models: ${inst.models.length > 0 ? inst.models.join(', ') : 'None listed'}`);
      console.log('');
    });
    
    return qualified;
  }

  /**
   * Complete workflow: Buy and prepare for AI consumption
   */
  async buyForAIConsumption(marketId, instrumentId, quantity) {
    console.log('\n=== Buy and Prepare for AI Consumption ===\n');
    
    // Step 1: Check if we have enough in consumption already
    const consumption = await this.client.getConsumptionInstruments();
    const existingInst = consumption.find(i => i.instrument_id === instrumentId);
    
    if (existingInst && existingInst.tradeable_amount >= quantity) {
      console.log(`✓ Already have ${existingInst.tradeable_amount} units in consumption`);
      return;
    }
    
    // Step 2: Check trading balance
    const tradingAccounts = await this.client.getTradingAccounts();
    const tradingAcct = tradingAccounts.find(a => a.instrument_id === instrumentId);
    
    const availableBalance = tradingAcct ? parseInt(tradingAcct.available_balance) : 0;
    
    if (availableBalance < quantity) {
      console.log(`Insufficient trading balance. Need ${quantity}, have ${availableBalance}`);
      console.log('Placing buy order...');
      
      // Get current market price
      const ticker = await this.client.getTicker(marketId);
      const buyPrice = parseFloat(ticker.lowest_ask) || parseFloat(ticker.last_price);
      
      // Place limit order at current ask
      const order = await this.client.placeOrder({
        market_id: marketId,
        side: 'buy',
        type: 'limit',
        quantity: quantity.toString(),
        price: buyPrice.toFixed(2),
        time_in_force: 'gtc'
      });
      
      console.log(`✓ Order placed: ${order.order_id}`);
      console.log('  Waiting for fill...');
      
      // In production, use WebSocket to wait for fill
      // For demo, we'll just wait a bit
      await new Promise(resolve => setTimeout(resolve, 5000));
    } else {
      console.log(`✓ Sufficient trading balance: ${availableBalance} units`);
    }
    
    // Step 3: Transfer to consumption
    console.log(`Transferring ${quantity} units to consumption...`);
    
    const transfer = await this.client.transferToConsumption(instrumentId, quantity);
    console.log(`✓ Transfer initiated: ${transfer.transfer_id}`);
    
    // Step 4: Confirm transfer
    console.log('Waiting for transfer completion...');
    
    let completed = false;
    let attempts = 0;
    
    while (!completed && attempts < 10) {
      await new Promise(resolve => setTimeout(resolve, 2000));
      
      const history = await this.client.getTransferHistory(null, 100);
      const transferRecord = history.find(t => t.transfer_id === transfer.transfer_id);
      
      if (transferRecord) {
        console.log('✓ Transfer completed');
        completed = true;
      }
      
      attempts++;
    }
    
    if (!completed) {
      console.log('⚠ Transfer taking longer than expected. Check transfer history.');
    }
    
    console.log('\n✓ Units ready for AI consumption!');
  }

  /**
   * Monitor market activity
   */
  async monitorMarket(marketId, durationMinutes = 5) {
    console.log(`\n=== Monitoring ${marketId} for ${durationMinutes} minutes ===\n`);
    
    const startTime = Date.now();
    const endTime = startTime + (durationMinutes * 60 * 1000);
    
    let tradeCount = 0;
    let totalVolume = 0;
    const priceHistory = [];
    
    while (Date.now() < endTime) {
      try {
        // Get recent trades
        const trades = await this.client.getPublicTrades(marketId, 5);
        
        if (trades.length > 0) {
          const latestTrade = trades[0];
          const price = parseFloat(latestTrade.price);
          priceHistory.push(price);
          
          console.log(`Trade: ${latestTrade.quantity} @ $${latestTrade.price} (${latestTrade.side})`);
          
          tradeCount++;
          totalVolume += parseInt(latestTrade.quantity);
        }
        
        // Get current order book
        const orderBook = await this.client.getOrderBook(marketId, 5);
        
        if (orderBook.bids.length > 0 && orderBook.asks.length > 0) {
          const [bidPrice, bidQty] = orderBook.bids[0];
          const [askPrice, askQty] = orderBook.asks[0];
          const spread = parseFloat(askPrice) - parseFloat(bidPrice);
          const spreadPct = (spread / parseFloat(bidPrice)) * 100;
          
          console.log(`  Bid: $${bidPrice} (${bidQty}) | Ask: $${askPrice} (${askQty}) | Spread: ${spreadPct.toFixed(3)}%`);
        }
        
        // Wait before next check
        await new Promise(resolve => setTimeout(resolve, 10000)); // 10 seconds
        
      } catch (error) {
        console.error(`Error during monitoring: ${error.message}`);
      }
    }
    
    // Summary
    console.log('\n=== Monitoring Summary ===');
    console.log(`Duration: ${durationMinutes} minutes`);
    console.log(`Trades Observed: ${tradeCount}`);
    console.log(`Total Volume: ${totalVolume} units`);
    
    if (priceHistory.length > 1) {
      const minPrice = Math.min(...priceHistory);
      const maxPrice = Math.max(...priceHistory);
      const avgPrice = priceHistory.reduce((a, b) => a + b, 0) / priceHistory.length;
      
      console.log(`Price Range: $${minPrice.toFixed(2)} - $${maxPrice.toFixed(2)}`);
      console.log(`Average Price: $${avgPrice.toFixed(2)}`);
    }
  }

  /**
   * Backtest a simple trading strategy
   */
  async backtestStrategy(marketId, days = 30) {
    console.log(`\n=== Backtesting Strategy on ${marketId} ===\n`);
    
    // Get historical data
    const now = Math.floor(Date.now() / 1000);
    const start = now - (days * 24 * 60 * 60);
    
    const candles = await this.client.getPriceHistory(marketId, '1h', start, now);
    
    console.log(`Loaded ${candles.length} hourly candles`);
    
    // Simple SMA crossover strategy
    let position = null;
    let cash = 10000;
    const trades = [];
    
    const calculateSMA = (data, period) => {
      const closes = data.map(c => parseFloat(c.close));
      return closes.reduce((a, b) => a + b, 0) / period;
    };
    
    for (let i = 20; i < candles.length; i++) {
      const currentPrice = parseFloat(candles[i].close);
      const sma20 = calculateSMA(candles.slice(i - 20, i), 20);
      
      // Buy signal: price crosses above SMA
      if (!position && currentPrice > sma20) {
        position = {
          entry: currentPrice,
          quantity: cash / currentPrice
        };
        cash = 0;
        trades.push({
          type: 'buy',
          price: currentPrice,
          quantity: position.quantity,
          time: candles[i].time
        });
        
        console.log(`BUY: ${position.quantity.toFixed(2)} units @ $${currentPrice.toFixed(2)}`);
      }
      // Sell signal: price crosses below SMA
      else if (position && currentPrice < sma20) {
        cash = position.quantity * currentPrice;
        trades.push({
          type: 'sell',
          price: currentPrice,
          quantity: position.quantity,
          time: candles[i].time
        });
        
        const pnl = cash - (position.quantity * position.entry);
        const pnlPct = (pnl / (position.quantity * position.entry)) * 100;
        
        console.log(`SELL: ${position.quantity.toFixed(2)} units @ $${currentPrice.toFixed(2)}`);
        console.log(`  P&L: $${pnl.toFixed(2)} (${pnlPct.toFixed(2)}%)`);
        
        position = null;
      }
    }
    
    // Close position if still open
    if (position) {
      const lastPrice = parseFloat(candles[candles.length - 1].close);
      cash = position.quantity * lastPrice;
      console.log(`CLOSE: ${position.quantity.toFixed(2)} units @ $${lastPrice.toFixed(2)}`);
    }
    
    // Results
    const finalValue = cash;
    const totalReturn = ((finalValue - 10000) / 10000) * 100;
    
    console.log('\n=== Backtest Results ===');
    console.log(`Initial Capital: $10,000.00`);
    console.log(`Final Value: $${finalValue.toFixed(2)}`);
    console.log(`Total Return: ${totalReturn.toFixed(2)}%`);
    console.log(`Number of Trades: ${trades.length}`);
    console.log(`Win Rate: ${this.calculateWinRate(trades).toFixed(1)}%`);
  }

  calculateWinRate(trades) {
    let wins = 0;
    let totalTrades = 0;
    
    for (let i = 0; i < trades.length - 1; i += 2) {
      if (trades[i].type === 'buy' && trades[i + 1].type === 'sell') {
        if (trades[i + 1].price > trades[i].price) {
          wins++;
        }
        totalTrades++;
      }
    }
    
    return totalTrades > 0 ? (wins / totalTrades) * 100 : 0;
  }

  /**
   * Smart instrument selector
   */
  async selectBestInstrument(requirements) {
    console.log('\n=== Finding Best Instrument ===\n');
    console.log('Requirements:', requirements);
    
    const instruments = await this.client.listInstruments();
    const aiInstruments = instruments.filter(i => i.instrument_type === 'ai_commodity');
    
    const candidates = [];
    
    for (const inst of aiInstruments) {
      const details = await this.client.getInstrument(inst.instrument_id);
      
      if (!details.ai_specs) continue;
      
      // Check requirements
      let score = 0;
      let qualifies = true;
      
      if (requirements.minContextWindow) {
        if (details.ai_specs.context_window >= requirements.minContextWindow) {
          score += 10;
        } else {
          qualifies = false;
        }
      }
      
      if (requirements.minThroughput) {
        if (details.ai_specs.token_throughput >= requirements.minThroughput) {
          score += 10;
        } else {
          qualifies = false;
        }
      }
      
      if (requirements.serviceType) {
        if (details.basic_info.service_type === requirements.serviceType) {
          score += 5;
        } else {
          qualifies = false;
        }
      }
      
      if (requirements.maxPrice && details.last_trade_price) {
        if (parseFloat(details.last_trade_price) <= requirements.maxPrice) {
          score += 10;
        } else {
          qualifies = false;
        }
      }
      
      if (qualifies) {
        candidates.push({
          ...details,
          score
        });
      }
    }
    
    // Sort by score
    candidates.sort((a, b) => b.score - a.score);
    
    console.log(`\nFound ${candidates.length} qualifying instruments:\n`);
    
    candidates.forEach((inst, idx) => {
      console.log(`${idx + 1}. ${inst.symbol}`);
      console.log(`   Context: ${inst.ai_specs.context_window.toLocaleString()} tokens`);
      console.log(`   Throughput: ${inst.ai_specs.token_throughput} t/s`);
      console.log(`   Price: ${inst.last_trade_price ? '$' + inst.last_trade_price : 'N/A'}`);
      console.log(`   Score: ${inst.score}/35`);
      console.log('');
    });
    
    return candidates.length > 0 ? candidates[0] : null;
  }
}

// Example usage
if (require.main === module) {
  const fs = require('fs');
  
  const privateKey = fs.readFileSync('./ed25519.key', 'utf8').trim();
  const publicKey = fs.readFileSync('./ed25519_pub.der', 'utf8').trim();
  
  const client = new AdvancedGridClient(privateKey, publicKey);
  
  (async () => {
    try {
      // Get portfolio overview
      await client.getPortfolioOverview();
      
      // Analyze a market
      const markets = await client.client.getMarkets();
      if (markets.length > 0) {
        await client.analyzeMarket(markets[0].market_id, 30);
        
        // Backtest strategy
        await client.backtestStrategy(markets[0].market_id, 30);
      }
      
      // Find best AI model
      const best = await client.selectBestInstrument({
        minContextWindow: 100000,
        minThroughput: 30,
        serviceType: 'text_generation',
        maxPrice: 100
      });
      
      if (best) {
        console.log(`\nRecommended: ${best.symbol}`);
      }
      
    } catch (error) {
      console.error('Error:', error.message);
    }
  })();
}

module.exports = { AdvancedGridClient };


